File: spark-java/spark-datasource/spark-ds-avro/spark-avro-config/src/main/java/com/mahfooz/config/pushdown/PredicatePushdown.java
package com.mahfooz.config.pushdown;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;

public class PredicatePushdown {
    public static void main(String[] args) {
        SparkSession spark = SparkSession.builder()
                .appName("PredicatePushdown")
                .master("local")
                .getOrCreate();

        String dataHome = System.getenv("DATA_HOME");
        String inputPath = String.format("%s/FileData/Avro/schema/users.avro", dataHome);

        // Read Avro data into a DataFrame
        Dataset<Row> avroDataFrame = spark.read()
                .format("avro")
                .load(inputPath);

        // Perform filtering after data has been loaded
        Dataset<Row> filteredDataFrame = avroDataFrame.filter("age > 25");

        // Show the content of the filtered DataFrame
        filteredDataFrame.show();

        filteredDataFrame.explain();

        // Stop SparkSession
        spark.stop();
    }
}

File: spark-scala/spark-datasources/spark-orc/src/main/scala/com/mahfooz/spark/orc/pushdown/predicate/PredicatePushdown.scala
package com.mahfooz.spark.orc.pushdown.predicate

import com.mahfooz.spark.orc.model.{Contact, Person}
import org.apache.spark.sql.{SaveMode, SparkSession}

object PredicatePushdown {

  val dataHome = sys.env.getOrElse("DATA_HOME","data")+"\\spark\\datasource\\orc"

  def main(args: Array[String]): Unit = {
    val spark = SparkSession.builder()
      .appName("predicate-pushdown")
      .master("local[*]")
      .getOrCreate()

    import spark.implicits._
    val records = (1 to 100).map { i =>
      Person(s"name_$i", i, (0 to 1).map { m => Contact(s"contact_$m", s"phone_$m") })
    }

    val dataFile = s"${dataHome}\\peoplePartitioned"
    records.toDF().write.format("orc")
      .mode(SaveMode.Overwrite)
      .partitionBy("age")
      .save(dataFile)

    spark.sql("set spark.sql.orc.filterPushdown=true")

    val people = spark.read.format("orc").load(dataFile)
    people.filter(people("age") < 15).select("name").explain(true)
  }

}
